### PR TITLE
Fixes #396: Restructure of applicatioon

### LINKED ISSUE #396: New data structure and actions for the Queries in the application
**Parent Issue #395**

The Data Structure should be implemented to accommodate the attributes that define a query in general.
This must also be associated with the Actions which update the state represented by the data structure.
In general the structure must contain minimum of `queryString` and a `FilterList`.

### PR DESCRIPTION
**Pull request by hemantjadon**
**Changes proposed in this pull request**
* This PR introduces the new query structure.
* Also includes the DeepFreeze ( Testing for Immutable state in Application)
*  This is a still a work in progress
* The tests might be failing
* The coverage will be going down
* Will be working on improving this

**Link to live demo (if appropriate):** 

https://hemantjadon.github.io/loklak_search/

**Closes #396 **

**Breaking Changes**
* This breaks currently the image rendering in cards, (This is quick fix, stopped due to the code was currently mutating the state)
* the query of type `followers: ...` these queries are currently not processed.

### CODE REVIEW

**Comment by hemantjadon** [comment]
@Achint08 @mariobehling Please Review,
@SKrPl @singhpratyush @vibhcool Please test with the test link and mention the areas where the application fails.
This is a big step in change and we will require a lot of testing.

**Comment by vibhcool** [comment]
@hemantjadon I did some queries like `fossasia`, `car`, `loklak`. I didn't get result for any.

**Comment by hemantjadon** [comment]
@vibhcool you would have to enable the site to allow `http` request, I mean the gh-pages site is on https and out API is currently on http, you should be seeing a red shield on the URL bar, if you are using chrome, on the right, and if you are using firefox, click on **Load Unsafe Scripts**

File: src/app/effects/api-suggest.effects.ts [hunk 1]
```diff
@@ -36,18 +36,18 @@ export class SuggestEffects {
 	@Effect()
 	suggest$: Observable<Action>
 		= this.actions$
-					.ofType(suggestAction.ActionTypes.SUGGEST)
-					.debounceTime(200)
+					.ofType(suggestAction.ActionTypes.SUGGEST_QUERY)
+					.debounceTime(400)
```
Comment (Achint08) [inline src/app/effects/api-suggest.effects.ts]: Decrease the debounceTime for suggest effects to make it more reactive.

File: src/app/reducers/query.ts [hunk 2]
```diff
@@ -0,0 +1,150 @@
...
+			const isFollowerQuery: boolean = (followersRegExp.exec(query)) ? true : false;
+
+			return Object.assign({}, state, {
+				query: {
+					...state.query,
+					displayString: query,
+					from: isFromQuery,
+					follower: isFollowerQuery
+				},
+				showUserFeed: isFollowerQuery
```
Comment (Achint08) [inline src/app/reducers/query.ts]: Not defined in the state, I guess.

File: src/app/effects/api-suggest.effects.ts (hunk 1 above)
Comment (hemantjadon) [inline src/app/effects/api-suggest.effects.ts]: I guess `300` will be optimal for this what do you say?

File: src/app/reducers/query.ts (hunk 2 above)
Comment (hemantjadon) [inline src/app/reducers/query.ts]: Yeah will remove it, this was really not a decent way to handle the `followers:<>` query, what was implemented earlier. I have something else in mind, will share  when this PR is done, don't want to make too many changes now, already there are 30+ modified files :/

Comment (Achint08) [inline src/app/reducers/query.ts]: Yeah. Good.  💯

File: src/app/effects/api-suggest.effects.ts (hunk 1 above)
Comment (Achint08) [inline src/app/effects/api-suggest.effects.ts]: Yeah it would be fine.

**Comment by hemantjadon** [comment]
@mariobehling Please review :)

**Comment by mariobehling** [comment]
@hemantjadon Thanks. Good job.

Let's make the review and merge process into a team collaboration thing. Please work with at least two collaborators to review and approve your PRs before mentors can merge it.

**Pull request merged**